unit/java_bytecode/generate_concrete_generic_type/generate_outer_inner.cpp

```diff
@@ -0,0 +1,367 @@
+      {
+        REQUIRE(new_symbol_table.has_symbol(expected_prefix));
+        const symbolt &expected_symbol =
+          new_symbol_table.lookup_ref(expected_prefix);
+
+        const struct_typet class_struct = to_struct_type(expected_symbol.type);
+
+        THEN("It has field f1 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f1");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::java.lang.Integer"));
+        }
+
+        THEN("It has field f2 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f2");
+          require_type::require_pointer(
+            field.type(),
+            symbol_typet("java::generic_outer_inner$GenericOuter$Inner"));
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_argument_kindt::Inst,
+              "java::java.lang.Integer"}});
+        }
+
+        THEN("It has field f3 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f3");
+          require_type::require_pointer(
+            field.type(),
+            symbol_typet(
+              "java::generic_outer_inner$GenericOuter$GenericInner"));
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_argument_kindt::Inst,
+              "java::java.lang.Integer"},
+             {require_type::type_argument_kindt::Inst,
+              "java::java.lang.String"}});
+        }
+      }
+    }
+
+    WHEN("Its field t2 is specialised")
+    {
+      generic_utils::specialise_generic_from_component(
+        class_prefix, "t2", new_symbol_table);
+      const std::string &expected_prefix =
+        class_prefix + "$GenericOuter<Integer>$Inner";
+      THEN("The corresponding specialised class symbol is created")
+      {
+        REQUIRE(new_symbol_table.has_symbol(expected_prefix));
+        const symbolt &expected_symbol =
+          new_symbol_table.lookup_ref(expected_prefix);
+
+        const struct_typet class_struct = to_struct_type(expected_symbol.type);
+
+        THEN("It has field f1 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f1");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::java.lang.Integer"));
+        }
+
+        THEN("It has field f2 of correct type")
+        {
+          const struct_union_typet::componentt &field =
+            require_type::require_component(class_struct, "f2");
+          require_type::require_pointer(
+            field.type(), symbol_typet("java::java.lang.Integer"));
+        }
```

Comment:
I think we should also check that there is a component called `this$0` which is the inner classes reference to the outer class. I think is is a struct-y type (rather than an a pointer) and as such should itself be specialised (i.e. should have a component called `f1` of type `Integer`. I fear this won't work - I'm happy for this to be raised as a bug as I don't think it is worse than the inheritance stuff.

Reply:
This check should be added to all inner classes as well

Reply:
I created a new task for recognizing pointers to generic outer classes as java generic types TG-1826. This should in turn result in specializing the `this$*` fields correctly. I added a TODO in the test with the link to the task.